datamade/la-metro-councilmatic
# Add forms to update bio and headshot

## Overview
As per issue #909 
> Add a form to the person detail page that allows logged in users to upload a headshot and/or input a bio to the person detail page.


### Checklist


### Demo
The user is logged in and has fields to edit the bio and headshot:
![image](https://user-images.githubusercontent.com/114717958/208190397-a5e1e97a-184a-47a4-9fca-a8c723345ba9.png)
<br/><br/>
After separately submitting a new image url like [this stock photo](https://images.unsplash.com/photo-1573496358961-3c82861ab8f4?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=688&q=80) and some lorem ipsum text into the bio:
![image](https://user-images.githubusercontent.com/114717958/208190805-d8147331-99b6-4c3e-b80e-478a16f10ccf.png)

### Notes

- I added some form validation into forms.py and PersonHeadshotForm now raises a few different errors on the page for different exceptions.
- I tried doing something similar to PersonBioForm but it seemed like no matter what adjustments I made, the form wasn't running the clean_bio method for validation. I put in print statements and that didn't show up, then I tried having it return the text if there was actually nothing there, but that didn't have any effect either. What I had was:
```
def clean_bio(self):
        bio_text = self.cleaned_data["biography"]

        if bio_text:
            return bio_text
        else:
            raise forms.ValidationError("The new bio cannot be empty. Please enter a biography.")
```
- On the plus side it looks like if someone were to remove the required attribute from the bio form, something in the backend actually isn’t updating the bio if it’s empty or just whitespace, and doing that on its own. Not sure how though.

## Testing Instructions

 * navigate to the board of directors page and select a person
 * the forms should not be visible until logging in
 * log in via the admin or otherwise, and navigate back to a person
 * enter a url into the headshot form and submit
   * if the url was valid, the headshot should change
   * else, an error is displayed
 - enter text into the bio form and submit
   * if text was entered, the bio should change
   * if only whitespace was entered, the bio stays the same

Closes #909
